### PR TITLE
feat: enable setting replicaSoftAntiAffinity per volume

### PR DESCRIPTION
**Pull request by ChanYiLin**
ref: [longhorn/longhorn#5358](https://github.com/longhorn/longhorn/issues/5358)

### CODE REVIEW

**Comment by innobead** [comment]
@ChanYiLin Please also add zone affinity as well. `ReplicaSoftAntiAffinity` and `ReplicaZoneSoftAntiAffinity`.

File: api/model.go [hunk 1]
```diff
@@ -937,9 +957,21 @@ func volumeSchema(volume *client.Schema) {
 	unmapMarkSnapChainRemoved := volume.ResourceFields["unmapMarkSnapChainRemoved"]
 	unmapMarkSnapChainRemoved.Required = true
 	unmapMarkSnapChainRemoved.Create = true
-	unmapMarkSnapChainRemoved.Default = false
+	unmapMarkSnapChainRemoved.Default = longhorn.UnmapMarkSnapChainRemovedIgnored
```
Comment (ChanYiLin) [inline api/model.go:960]: I believe this should be `longhorn.UnmapMarkSnapChainRemovedIgnored` string value instead of bool value `false`

**Comment by ChanYiLin** [comment]
344071214_492371553018167_8763100615334666306_n

volume now has new field `ReplicaSoftAntiAffinity` and `ReplicaZoneSoftAntiAffinity` in spec and can get from api

File: scheduler/replica_scheduler.go [hunk 2]
```diff
@@ -146,17 +146,31 @@ func getNodesWithEvictingReplicas(replicas map[string]*longhorn.Replica, nodeInf
...
 	if err != nil {
 		logrus.Errorf("error getting replica soft anti-affinity setting: %v", err)
 	}
+	if volume.Spec.ReplicaSoftAntiAffinity != longhorn.ReplicaSoftAntiAffinityDefault &&
+		volume.Spec.ReplicaSoftAntiAffinity != "" {
+		if volume.Spec.ReplicaSoftAntiAffinity == longhorn.ReplicaSoftAntiAffinityEnabled {
+			nodeSoftAntiAffinity = true
+		} else {
+			nodeSoftAntiAffinity = false
+		}
```
Comment (innobead) [inline scheduler/replica_scheduler.go:159]: Can be. Also, apply this change to `zoneSoftAntiAffinity`

```go
nodeSoftAntiAffinity = volume.Spec.ReplicaSoftAntiAffinity == longhorn.ReplicaSoftAntiAffinityEnabled
```

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, just a few feedbacks.

File: scheduler/replica_scheduler.go (hunk 2 above)
Comment (ChanYiLin) [reply scheduler/replica_scheduler.go]: Fixed, thanks!

File: k8s/pkg/apis/longhorn/v1beta2/volume.go [hunk 3]
```diff
@@ -216,6 +234,10 @@ type VolumeSpec struct {
 	// +optional
 	UnmapMarkSnapChainRemoved UnmapMarkSnapChainRemoved `json:"unmapMarkSnapChainRemoved"`
 	// +optional
+	ReplicaSoftAntiAffinity ReplicaSoftAntiAffinity `json:"replicaSoftAntiAffinity"`
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply k8s/pkg/apis/longhorn/v1beta2/volume.go:238]: Fixed, thanks!

File: manager/volume.go [hunk 4]
```diff
@@ -154,27 +154,29 @@ func (m *VolumeManager) Create(name string, spec *longhorn.VolumeSpec, recurring
 			Labels: labels,
 		},
 		Spec: longhorn.VolumeSpec{
```
[the remark this answers is not in the document]
Comment (ChanYiLin) [reply manager/volume.go:156]: Agree.

[the remark this answers is not in the document]
Comment (innobead) [reply manager/volume.go:156]: Let's improve this later.

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, buneed to regeneate crds.yaml due to new description added to the new fileds. 

After this, we can merge this PR.

**Comment by ChanYiLin** [comment]
> LGTM, buneed to regeneate crds.yaml due to new description added to the new fileds.
> 
> After this, we can merge this PR.

Hi @innobead, thanks for the reminder, fixed, thanks!

**Review by innobead — APPROVED** [review]

**Pull request merged**